src/ess/dream/diagnostics.py

```diff
@@ -0,0 +1,228 @@
+        return fig
+
+    @staticmethod
+    def _prepare_data(dg: sc.DataGroup) -> sc.DataGroup:
+        return sc.DataGroup(
+            {
+                name: bank.transpose(
+                    [dim for dim in _STARTING_DIMS if dim in bank.dims]
+                )
+                for name, bank in dg.items()
+            }
+        )
+
+
+class _DimensionSelector(ipw.VBox):
+    def __init__(self, dims: tuple[str, ...]) -> None:
+        self._value_observers = []
+
+        h_buttons, v_buttons = self._make_buttons(dims, *self._default_dims(dims))
+
+        self._horizontal_selector_box = ipw.HBox([h_buttons])
+        self._vertical_selector_box = ipw.HBox([v_buttons])
+        super().__init__(
+            [
+                ipw.HBox([ipw.Label('X'), self._horizontal_selector_box]),
+                ipw.HBox([ipw.Label('Y'), self._vertical_selector_box]),
+            ]
+        )
+
+    def _make_buttons(
+        self, dims: tuple[str, ...], h_dim: str, v_dim: str
+    ) -> tuple[ipw.ToggleButtons, ipw.ToggleButtons]:
+        style = {'button_width': '10em'}
+        h_buttons = ipw.ToggleButtons(options=dims, value=h_dim, style=style)
+        v_buttons = ipw.ToggleButtons(options=dims, value=v_dim, style=style)
+        h_buttons.observe(self._observe_values, names='value')
+        v_buttons.observe(self._observe_values, names='value')
+        return h_buttons, v_buttons
+
+    @property
+    def _horizontal_buttons(self) -> ipw.ToggleButtons:
+        return self._horizontal_selector_box.children[0]
+
+    @property
+    def _vertical_buttons(self) -> ipw.ToggleButtons:
+        return self._vertical_selector_box.children[0]
+
+    def set_dims(self, new_dims: tuple[str, ...]) -> None:
+        self._horizontal_buttons.unobserve_all(name='value')
+        self._vertical_buttons.unobserve_all(name='value')
+        # self._forward_messages = False
+
+        old_h = self._horizontal_buttons.value
+        old_v = self._vertical_buttons.value
+        default_h, default_v = self._default_dims(new_dims)
+        new_h = old_h if old_h in new_dims else default_h
+        new_v = old_v if old_v in new_dims else default_v
+
+        h_buttons, v_buttons = self._make_buttons(new_dims, new_h, new_v)
+        self._horizontal_selector_box.children = [h_buttons]
+        self._vertical_selector_box.children = [v_buttons]
+
+    @staticmethod
+    def _default_dims(dims: tuple[str, ...]) -> tuple[str, str]:
+        return dims[math.ceil(len(dims) / 2)], dims[0]
+
+    @property
+    def value(self):
+        return {
+            'horizontal': self._horizontal_buttons.value,
+            'vertical': self._vertical_buttons.value,
+        }
+
+    def observe(self, handler, names: str | list[str], *args, **kwargs):
+        if names != 'value':
+            super().observe(handler, names, *args, **kwargs)
+        else:
+            self._value_observers.append(handler)
```

Comment:
Not sure I understand the mechanism with the `value_observers`. Maybe you can explain to me?

Reply:
This is about updating only once when dimensions are swapped between x and y. I tried to find a way to stop widgets from emitting messages. But I couldn't get this to work. So instead, I put a custom layer in between so that I can control what messages make it to the user.

All observers with `names='value` are registered in `self._value_observers`. And the actual handler for the toggle buttons is `self._observe_values`. This function calls the user handlers when appropriate.

Reply:
How about having a lock?
Add a `self._lock = False`.

Then in the function that gets called on update:
```Py
def update(change):
    if self._lock:
        return
    if other.value == clicked.value:
        self._lock = True
        other.value = change["old"]
        self._lock = False
    ...
```
Would that work? It would simplify a lot of the logic I think?

Reply:
I don't think this would work. This would prevent updating within `_DimensionSelector` while the lock is held. But the observers of the button that control the actual figure would still get called on `other.value = change["old"]`.

Reply:
Seems to work. Thanks!

I don't think this solution is valid for multi-threaded applications. But I am not 100% confident about my old implementation either. And it should be fine because ipywidgets does not spawn threads, right?